events/internal/equeue.h

```diff
@@ -26,11 +24,14 @@ extern "C" {
 #endif
 
 // Platform specific files
-#include "equeue/equeue_platform.h"
+#include "events/internal/equeue_platform.h"
 
 #include <stddef.h>
 #include <stdint.h>
 
+/** \ingroup events */
```

Comment:
I think that should be a public header. I know it's sort of underlying concept, but in some situations (eg bare metal) people may not want to use C++. We also need to do something smart about the both README files, I would say merge them into one. Starting with the C++ description and then introduce concept of underlying C APIs and that they can be used in bare metal c only projects.

Reply:
@geky fyi

Reply:
Well, opinions aside, this just made all the subtrees quite a bit harder to maintain :(

Are we doing the same for littlefs and lwip?

Reply:
I can't even imagine sub-trees being any harder to maintain than they already were when I last tried to use sub trees for dependency management.

Reply:
Oh really? I'm actually a big fan of subtrees :) Though as I understand the quality of the command has changed quite a bit over the last half-decade.

It works really well at separating version control from other programmer debates such as monorepo/multirepo, build/dependency systems.

It can get complicated, but one of the best things it has going for it is that, all the effort is on the developer making the subtree merge. After that the user never needs to know (unlike submodules) and even other developers can work on merged directory as though it was part of the repo.

Every so often I'll come back through and create another subtree merge pr to bring the code up to date with what's outside the tree (though I've been slow because of littlefs v2 stuff, sorry about that): https://github.com/ARMmbed/mbed-os/pull/7713

Subtrees really seem like a realization of the idea of truly distributed version control. Because of it repos can exist as copies in a bunch of different places with their own quirks and patches.

Reply:
After thinking about this, I think it's time that equeue and Mbed OS split and go their separate ways.

In general, the feedback I've gotten on equeue in Mbed OS has been negative. I think at this point it's overstayed its welcome.

Not that Mbed OS should go without an event queue, just that Mbed OS's event queue should be revamped and get a new owner (@kjbracey-arm, nanostack-eventloop? 🙂).

equeue originally started as a standalone, OS-agnostic project over here: https://github.com/geky/equeue. I still maintain it, though it's relatively quiet. The goal was to have a small, reliable, and useful standalone tool for scheduling on microcontrollers, because, from my limited experience, this is all you need 90% of the time. And by being OS-agnostic, you can write an application that is more or less portable across different environments.

You may notice this doesn't really match the idea behind Mbed OS, like, at all. And I think that's creating some conflicting design decisions.

It's in equeue's best interest to keep the API as minimal as possible, as this reduces the surface area for bugs and quirky interactions. This is why I push back so hard on new features in an effort to stop feature-creep and keep the scope limited. But I'm just creating problems if your only platform is Mbed OS and you're trying to solve some Mbed OS-centric issues:
- event priorities https://github.com/ARMmbed/mbed-os/issues/10263
- static allocation https://github.com/ARMmbed/mbed-os/issues/9172
- cancel to return success https://github.com/ARMmbed/mbed-os/pull/10750

Going the other way, Mbed OS doesn't need a number of things in equeue:
- FreeRTOS/Window/etc support
- Backgrounding for when you only have hardware timers
- Chaining makes less sense when you have threads 

So my conclusion is that these two projects should diverge.

A graceful way to do this would be to rename equeue -> mbed_events or something similar in Mbed OS so it matches the name of C++ API. @bulislaw I think you're right we should merge the READMEs and I think that extends to the directory structure as well. Basically abolish the separation that currently exists because equeue is also a standalone repo. I know it's not great, but a name change would be very helpful as this way equeue can also be brought in as a separate library without name conflicts.

Then, once time permits, we can introduce nanostack-eventloop and mbed-events to each other and see what happens.

That's my proposal, if we're moving around a lot of files it would be awfully convenient to do this now. I can give you a sed script or add a commit if you'd like.

Thoughts?

@bulislaw, @kjbracey-arm, @pan-, @c1728p9, @evedon

Reply:
Thiugh I have a preference for single threaded event queing over threading...

Your statements above feel right.

Mbed OS needs an event queing system that can be more flexible in it's interactions with other components of the OS and more actively maintained.

I'd say turn your statement above into it's own. Issue and go for it.

Reply:
@geky Let's not throw the baby out with the bathwater. There's some things the `EventQueue` does better than other event systems present in mbed-os tree :
* It's not a singleton and allows you to have several queue coexisting in a system. 
* Allocation system works great with C++ lambda
* Allocation is per queue; not global.  

Consolidation of a single event system would be great to reduce code size, maintenance effort and bug surface but, first, we need to gather requirements from all these systems (`EventQueue`, `nano stack event loop`, `WSF event queue`, ...) to make sure we're going in the right direction before jumping on a solution.

----

> You may notice this doesn't really match the idea behind Mbed OS, like, at all. And I think that's creating some conflicting design decisions.

How so ? Queuing event will **always** be relevant and if it can run on x86 it can help for integration and unit testing as well.

Reply:
This is a useful discussion but I am very reluctant to add scope to this already huge PR. So except for @bulislaw original comment on whether equeue should be regarded as a public or internal API, I'd like everything else to be dealt with separately, specifically: the README cleanup and the potential renaming of equeue to mbed_events.

Reply:
@evedon, good point. I'll move this to a separate issue.

My vote for equeue public vs internal is to leave it as internal for now. Are we worried this will break code?

Reply:
> Let's not throw the baby out with the bathwater. There's some things the EventQueue does better than other event systems present in mbed-os tree :

To be clear I'm not suggesting we remove EventQueue. Or even remove any features yet.

Just that we rename equeue -> mbed_events (or similar), remove separate between equeue and EventQueue where it makes sense, and let the projects naturally diverge as their features grow/shrink to match their requirements.

> How so ? Queuing event will always be relevant and if it can run on x86 it can help for integration and unit testing as well.

I was more referring specifically to the portability/OS-agnostic aspect of equeue.

You're right x86 support has been useful for testing, but IMO x86 testing should somehow be provided for all of Mbed OS. It's the only way to keep this shindig scalable.

FreeRTOS and Windows support have already been removed from the Mbed OS version because it makes less sense here. And there have been several cases of the terminology differences between equeue and the RTOS confusing users.

Reply:
I agree :D So how about for this PR we:
* Make queue public, as we need C API for bare metal builds
* Do not rename anything public so that we don't break compatibility
Going forward:
* @kjbracey-arm is looking at getting all the requirements from all the components so we can finaly have one queue to rule them all (mandatory https://xkcd.com/927/)
* @donatieng @pan- are looking at 2 priorities for the queue and at static allocations
For mbed 6 we will try to force the unification and possible changes to the existing APIs.

I'll catch up with y'all at some point to discuss the plans further.

Does it sound like a plan?

Reply:
👍 
@bulislaw @kjbracey-arm @donatieng @pan  
I am looking forward to the discussion of this topic further in the related public issue/issues

Reply:
I'm a bit confused. My conclusion was that there shouldn't be a single queue. Is there plans to host a version outside of mbed-os?

> @donatieng @pan- are looking at 2 priorities for the queue and at static allocations

FYI, I am also looking at static allocations, (though intentionally not priorities). Though a part that plan was to break API compatibility to introduce better error codes. Let me know if you want me to create a PR here as well though.